javascript/src/refactor.js

```diff
@@ -0,0 +1,648 @@
+        this.keep_comments = keep_comments;
+        this.behaviour = behaviour;
+        this.changed = false;
+        this.filename = filename;
+    }
+        
+    trueLiteral() { 
+        const vanillaLiteral = { // A Boolean Literal is parsed like this by recast
+            type: "Literal", 
+            value: true, 
+            raw: 'true'
+        };
+    
+        // This additional property marks that this literal was introduced by Piranha 
+        vanillaLiteral.createdByPiranha = true;
+    
+        return vanillaLiteral;
+    }
+    
+    falseLiteral() { 
+        const vanillaLiteral = { // A Boolean Literal is parsed like this by recast
+            type: "Literal", 
+            value: false, 
+            raw: 'false'
+        };
+    
+        // This additional property marks that this literal was introduced by Piranha 
+        vanillaLiteral.createdByPiranha = true;
+    
+        return vanillaLiteral;
+    }
+
+    getMethodHashMap(properties) { 
+        var methodHashMap = new Map();
+    
+        properties["methodProperties"].forEach((item) => {
+            methodHashMap.set(item["methodName"], item);
+        });
+    
+        return methodHashMap;
+    }
+
+    // Verify this is a literal introduced by Piranha. 
+    // We only refactor code associated with literals having this property and leave original code untouched
+    isPiranhaLiteral(node) {
+        return node.type === "Literal" && node.createdByPiranha !== undefined;
+    }
+
+    reduceLogicalExpression(literal, expression, operator) {
+        if (operator === "&&") {
+            if (literal.value === true) {
+                return expression;
+            }
+                
+            if (literal.value === false) {
+                return this.falseLiteral();
+            }
+        } else if (operator === "||") {
+            if (literal.value === true) {
+                return this.trueLiteral();
+            }
+                
+            if (literal.value === false) {
+                return expression;
+            }
+        }
+    }
+
+    // Check if a function with return statement at the end of function body is redundant, i.e returns piranhaLiteral
+    checkAndAddRedundantFunction(func, name, redundantFunctions) {
+        if (func.body === null || func.body.body == null) {
+            return false;
+        }
+
+        const returnIndex = func.body.body.length - 1; 
+
+        if (returnIndex < 0) // If returnIndex == -1 when function body is empty
+            return false;
+        else {
+            var returnNode = func.body.body[returnIndex];
+
+            if (returnNode.type === "ReturnStatement" && 
+                returnNode.argument !== null && 
+                this.isPiranhaLiteral(returnNode.argument) && 
+                typeof(returnNode.argument.value) === "boolean") { 
+                    redundantFunctions[name] = returnNode.argument.value; 
+            }
```

Comment:
What happens here with code like:

```
foo(...) {
    if(flagOne) {
        return X;
    }

    if(flagTwo) {
        return Y;
    }
}
```

When simplifying `flagTwo`? I feel it will ignore `flagOne` and simplify that to a redundant function, unless I am missing something...

Reply:
Actually. I see below that only functions with a single return statement will get passed to this method, which solves that issue. Maybe worth a comment reminding the reader of that here 🙂

Reply:
Done 7d274de83d26b96acb5de427115b208cfc7db26c